colcon_core/topological_order.py

```diff
@@ -62,12 +62,12 @@ def topological_order_packages(
 
     # create ordered list of package decorators
     decorators = []
+    ordered_keys = [descriptor.name for descriptor in ordered.keys()]
     for descriptor, recursive_dependencies in ordered.items():
         decorator = PackageDecorator(descriptor)
         # reorder recursive dependencies according to the topological ordering
-        decorator.recursive_dependencies = [
-            desc.name for desc in ordered.keys()
```

Comment:
Wouldn't it be sufficient to remove the `.name` from `desc.name` here to achieve the same thing?

Reply:
In this spot, `desc` is a `PackageDescriptor`, where we want a `DependencyDescriptor`. They both have a `.name` attribute.

Reply:
The patched code has a different logic than before. Before  items from `recursive_dependencies` were being added only if they are present in `ordered.keys()` and the result is ordered in the same way as `ordered.keys()`. With the patch all items from `recursive_dependencies` are in the result. I would rather not have the patch change the semantic (however that can be achieved).

Reply:
If there were elements in `recursive_dependencies` not present in `ordered.keys()`, we would be seeing a `ValueError` thrown by `ordered_keys.index()`, but we're not. Something else must be removing the unknown dependencies.

I'll try to figure out where that's happening.

Reply:
Anyway you shouldn't rely on that to be the case in this code snippet. I would suggest to modify the patch to maintain the current logic while returning the descriptors instead of the name.

Reply:
Ah, I found it. `get_recursive_dependencies()` already makes that guarantee right here: https://github.com/colcon/colcon-core/blob/master/colcon_core/package_descriptor.py#L115

Do you still want it to check again, even though `recursive_dependencies` is already populated directly by `get_recursive_dependencies()`?

Reply:
Embedding this assumption in the logic add an unnecessary unobvious coupling. So I think the logic should be preserved since it is more defensive programming.

Reply:
Okay, I made the change. My tests show no measurable performance loss from the additional check.